cherry picks of some betterai BT changes

bmorel opened the pull request:
This PR cherry picks 3 commits allowing bots to go grab ammo and poison when they need to. Needs to be used by BT to have any impact on gameplay.

bmorel commented:
Note that this PR does **not** include a BT action that make bots reload, as I remember that this action required some heavier modifications to work reliably.

bmorel commented:
I removed the commit that implements `action resupply` as it's already possible to do that with the current API, using `changeGoal` and `moveToGoal` actions.

slipher commented on src/sgame/sg_bot_util.cpp:
```diff
@@ -365,7 +365,7 @@ float BotGetResupplyScore( gentity_t *self )
 	}
 	float timeDist = dist / GetMaximalSpeed( self );
-	return ( 1 + 5 * self->botMind->botSkill.level ) * ( 1 - percentAmmo ) / sqrt( timeDist );
+	return ( 1 + 5 * self->botMind->botSkill.level ) * ( needPoison ? 1 : ( 1 - percentAmmo ) ) / sqrt( timeDist );
```
The poison thing could be kind of broken for large maps; you may not be able to get to the human base before the poison runs out.

bmorel replied:
Good point.
But this is easy to counter in the BT itself by checking distance to BOOSTER before checking for the score. I believe this is something that should be done more often anyway, notably for the actions `heal` and `equip`.

IIRC the logic here is the same as in `heal` for the distance (or was when I initially implemented this).

slipher commented on src/sgame/sg_bot_parse.cpp:
```diff
@@ -148,9 +148,48 @@ static AIValue_t haveUpgrade( gentity_t *self, const AIValue_t *params )
+	playerState_t const& ps = self->client->ps;
+	weaponAttributes_t const* wpa = BG_Weapon( BG_PrimaryWeapon( ps.stats ) );
+
+	if ( wpa->maxAmmo == 0 && !wpa->infiniteAmmo )
```
The warning seems unnecessary as you already check these conditions in `Ratio` and handle them correctly. And it would print every frame here, so if you really want a warning it should be in the weapon parser code.

bmorel replied:
Fair. I'll remove this. I guess I was just going paranoid here, or maybe it's simply older than the implementation of `Ratio`. I don't remember history details behind this.

slipher reviewed (commented):
I see it's been simplified a bit by removing goon barb support. Reducing the scope of the resupply score to a single responsibility (only poison rather than poison + barbs) resolves my concern about getting poison not always being feasible. LGTM

bmorel commented:
> I see it's been simplified a bit by removing goon barb support. Reducing the scope of the resupply score to a single responsibility (only poison rather than poison + barbs) resolves my concern about getting poison not always being feasible.

No, there is still barb support: if the bot have no need for poison (because it already have poison), then it still might go (or rather stay) at booster for ammo, this is what `bool needAmmo = !( weapon->infiniteAmmo && weapon->maxAmmo == 0 );` is for: there is a need for ammo if either condition is met: no infinite ammo, or maxAmmo != 0. This means that melee attacks won't set needAmmo, but barbs (the only weapon which have regenerating ammo) will.

I think those commits represent almost all the resupply API I used in betterai, with the only exception of the `action reload` which triggers bugs in the memory system. Those bugs i've fixed too, but the changeset is *huge* and I would not consider it ready for merge. It works without known problems, but it does not completely fix bot's memory issues so I'm not happy with it and won't try to push that for now (but maybe I could extract few changes from there that fix some actual bugs in enemy detection? I'll think about it.).

bmorel commented:
I updated the comments as discussed on IRC, and added support for energy resupply near drills, as when I wrote this drills were not able to resupply energy weapons.

slipher commented on src/sgame/sg_bot_parse.cpp:
```diff
@@ -148,9 +148,39 @@ static AIValue_t haveUpgrade( gentity_t *self, const AIValue_t *params )
+}
+
+// Return emptyness ammo ratio of currently loaded clip
+// TODO: implement for non-primary weapon (such as barbs or future secondary weapons)
```
Oh there's two more comments here claiming it doesn't work for barbs.

slipher commented:
LGTM
